src/service/referral-code-service.ts

```diff
@@ -0,0 +1,306 @@
+
+        const dataValues = [...filter.values, take, skip];
+        const dataQuery = {
+            text: `SELECT id, name, type, valid_from, code, valid_to, instructions_url, project_group_id, user_id, created_at, updated_at, description, is_active
+                   FROM promo_referrals ${filter.whereClause}
+                   ORDER BY created_at DESC
+                       LIMIT $${filter.nextIndex} OFFSET $${filter.nextIndex + 1}`,
+            values: dataValues,
+        };
+
+        const dataResult = await dbClient.query(dataQuery);
+        const data = dataResult.rows.map(row => ReferralCodeDto.from(row));
+
+        const totalPages = take === 0 ? 0 : Math.ceil(totalItems / take);
+
+        return {
+            total_items: totalItems,
+            per_page: take,
+            current_page: pageNo,
+            total_pages: totalPages,
+            data,
+        };
+    }
+
+    async createReferralCode(projectGroupId: string, referralCode: ReferralCodeDto, userId: string): Promise<ReferralCodeDto> {
+        if (!userId) {
+            throw new HttpException(400, "User context missing for creating referral code");
+        }
+        const normalized = this.normalizeReferralCodeInput(referralCode);
+        normalized.project_group_id = projectGroupId;
+        normalized.user_id = userId;
+
+        const existingCode = await this.checkReferralCodeExists({ code: normalized.code });
+        if (existingCode) {
+            throw new HttpException(402, "Code already exists");
+        }
+
+        const query = {
+            text: `INSERT INTO promo_referrals
+                   (name, type, valid_from, code, valid_to, instructions_url, project_group_id, user_id, description, is_active)
+                   VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, true)
+                       RETURNING id, name, type, valid_from, code, valid_to, instructions_url, project_group_id, user_id, created_at, updated_at, description, is_active`,
+            values: [
+                normalized.name,
+                normalized.type,
+                normalized.valid_from,
+                normalized.code,
+                normalized.valid_to ?? null,
+                normalized.instructions_url ?? null,
+                normalized.project_group_id,
+                normalized.user_id,
+                normalized.description ?? null,
+            ],
+        };
+
+        try {
+            const result = await dbClient.query(query);
+            return ReferralCodeDto.from(result.rows[0]);
+        }
+        catch (error) {
+            if (error instanceof UniqueKeyDbException) {
+                throw new DuplicateException(`Referral code '${normalized.code}' already exists.`);
+            }
+            throw error;
+        }
+    }
+
+    async updateReferralCode(projectGroupId: string, codeId: string, referralCode: ReferralCodeDto, userId: string): Promise<ReferralCodeDto> {
```

Comment:
please rename `codeId` to `referralCodeId`. Otherwise, it gets mistaken as `shortcode`